### PR TITLE
(stream) signl4-sc initial release

### PR DESCRIPTION
**Pull request by Sims24**

### CODE REVIEW

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    -- checking mandatory parameters and setting a fail flag
+    if not self.sc_params:is_mandatory_config_set(mandatory_parameters, params) then
+      self.fail = true
+    end
+
+    -- overriding default parameters for this stream connector if the default values doesn't suit the basic needs
+    self.sc_params.params.accepted_categories = params.accepted_categories or "neb"
+    self.sc_params.params.accepted_elements = params.accepted_elements or "host_status,service_status"
+    self.sc_params.params.x_s4_service = params.x_s4_service or "Centreon"
+    self.sc_params.params.http_server_url = params.http_server_url or "https://events.pagerduty.com/v2/enqueue"
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:61]: not quite sure about that :D

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    self.sc_params.params.client = params.client or "Centreon Stream Connector"
+    self.sc_params.params.accepted_categories = params.accepted_categories or "neb"
+    self.sc_params.params.accepted_elements = params.accepted_elements or "host_status,service_status"
+    self.sc_params.params.pdy_source = params.pdy_source or nil
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:65]:
```suggestion
```

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+  --   "X-S4-Service": "Centreon",
+  --   "X-S4-AlertingScenario": "multi_ack", #NOT INCLUDED ATM
+  --   "X-S4-ExternalID": "HOSTALERT_666",
+  --   "X-S4-Status": "new"
+  -- }
+
+  function EventQueue:format_event_host()
+    self.sc_event.event.formated_event = {
+      "Event Type" = "HOST",
+      "Date" = self.sc_macros:replace_sc_macro(self.sc_event.event.last_check),
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:149]: do you mean 
https://github.com/centreon/centreon-stream-connector-scripts/blob/master/modules/docs/sc_macros.md#transform_date-method ?

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+      "X-S4-Service" = self.sc_params.params.x_s4_service,
+      "X-S4-ExternalID" = "HOSTALERT_" .. event.host_id, 
+      "X-S4-Status" = self.state_to_signlstatus_mapping[event.state]
+    }
+  end
+
+  function EventQueue:format_event_service()
+    self.sc_event.event.formated_event = {
+      "Event Type" = "SERVICE",
+      "Date" = self.sc_macros:replace_sc_macro(self.sc_event.event.last_check),
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:165]: do you mean 
https://github.com/centreon/centreon-stream-connector-scripts/blob/master/modules/docs/sc_macros.md#transform_date-method ?

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+  function EventQueue:send_data(data, element)
+    self.sc_logger:debug("[EventQueue:send_data]: Starting to send data")
+
+    -- write payload in the logfile for test purpose
+    if self.sc_params.params.send_data_test == 1 then
+      self.sc_logger:info("[send_data]: " .. broker.json_encode(data))
+      return true
+    end
+
+    local http_post_data = broker.json_encode(http_post_metadata)
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:208]: http_post_metadata doesn't exist

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    for _, raw_event in ipairs(data) do
+      http_post_data = http_post_data .. broker.json_encode(http_post_metadata) .. "\n" .. broker.json_encode(raw_event) .. "\n"
+    end
+
+    self.sc_logger:info("[EventQueue:send_data]: Going to send the following json " .. tostring(http_post_data))
+    self.sc_logger:info("[EventQueue:send_data]: Signl4 Server URL is: " .. tostring(self.sc_params.params.server_address) .. "/webhook/" .. tostring(self.sc_params.params.team_secret))
+
+  local http_response_body = ""
+  local http_request = curl.easy()
+    :setopt_url(self.sc_params.params.server_address .. "/_bulk")
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:218]: it looks like elastic api, is it normal ?

**Review by tanguyvda — CHANGES_REQUESTED** [review]

File: centreon-certified/signl4/signl4-events-apiv2.lua [hunk 7]
```diff
@@ -191,85 +189,77 @@ function EventQueue:add()
...
-      self.sc_logger:info("[EventQueue:send_data]: HTTP POST request successful: return code is " .. tostring(http_response_code))
-      retval = true
+  end
+  -- set proxy user configuration
+  if (self.sc_params.params.proxy_username ~= '') then
+    if (self.sc_params.params.proxy_password ~= '') then
+      http_request:setopt(curl.OPT_PROXYUSERPWD, self.sc_params.params.proxy_username .. ':' .. self.sc_params.params.proxy_password)
     else
-      self.sc_logger:error("[EventQueue:send_data]: HTTP POST request FAILED, return code is " .. tostring(http_response_code) .. ". Message is: " .. tostring(http_response_body))
+      broker_log:error("[EventQueue:send_data]: proxy_password parameter is not set but proxy_username is used")
```
Comment (tanguyvda) [inline centreon-certified/signl4/signl4-events-apiv2.lua:243]: why broker_log ?

**Review by tanguyvda — CHANGES_REQUESTED** [review]

File: centreon-certified/signl4/signl4-events-apiv2.lua (hunk 7 above)
Comment (Sims24) [reply centreon-certified/signl4/signl4-events-apiv2.lua]: parce-que-orangina

Should be ok now ;)

**Review by tanguyvda — APPROVED** [review]

**Pull request merged**